### PR TITLE
submission status callback (#211)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
Addresses #211.

```
from taipy.gui_core.GuiCoreLib import _GuiCore

from taipy.gui import Gui, State
from taipy import Config, Core, Submittable
import taipy as tp


def on_creation(state: State, id: str, action: str, payload: dict):
    print(f"on_creation(state: State, id: str, action: str, payload: dict) => {payload.get('label')}")


# Normal function used by Taipy
def d(nb):
    return nb * 2


def submission_change(state: State, submittable: Submittable, details: dict):
    print(f"submission_change(state, submittable: {submittable}, details: {details})")


i_cfg = Config.configure_data_node("input", default_data=21)
o_cfg = Config.configure_data_node("output")

task_cfg = Config.configure_task("double", d, i_cfg, o_cfg)

scenario_cfg = Config.configure_scenario("my_scenario", [task_cfg])

Config.export("demo-config_01.toml")
selected_scenario = None
none_scenario = None
selected_data_node = None
selected_jobs = None

if __name__ == "__main__":
    # Run of the Core
    Core().run()

    # Creation of the scenario and execution
    selected_scenario = tp.create_scenario(scenario_cfg)
    tp.submit(selected_scenario)

    md = """
<|{selected_scenario}|scenario_selector|>
<|{selected_scenario}|scenario|>
<|{selected_scenario}|scenario_dag|>
"""

    page = """
## Demo Core Gui
<|toggle|theme|>

<|layout|columns=1 1|

<|
### Scenario Selector
<|{selected_scenario}|scenario_selector|on_creation=on_creation|>
|>

<|
### DataNode Selector
<|{selected_data_node}|data_node_selector|>
|>

<|
### Visualizer
<|{selected_scenario}|scenario|on_submission_change=submission_change|>
|>

<|
### Selected DataNode
<|{'None' if selected_data_node is None else selected_data_node.get_label()}|>
<|{selected_data_node}|data_node|scenario={selected_scenario}|>
<|{none_scenario}|data_node|>
|>


### DAG
<|{selected_scenario}|scenario_dag|>

|>

### Job Selector
<|{selected_jobs}|job_selector|>


"""

    Gui(page).run()

```

### CODE REVIEW

File: src/taipy/gui_core/_context.py [hunk 1]
```diff
@@ -271,11 +298,110 @@ def submit_entity(self, state: State, id: str, action: str, payload: t.Dict[str,
...
+            if not callable(on_action_function):
+                return
+
+            new_status = self._get_submittable_status(jobs)
+            if current_status is not new_status:
+                # callback
+                self.gui._call_user_callback(
+                    client_id,
+                    on_action_function,
+                    [entity, {"submission_status": new_status.name, "job": job}],
```
Comment (FabienLelaquais) [inline src/taipy/gui_core/_context.py:413]: I think the first parameter should be... something else.
In the future, that will be a 'Submission'. Maybe we need to create a temporary wrapper class for that.

File: src/taipy/gui_core/_context.py [hunk 2]
```diff
@@ -271,11 +298,110 @@ def submit_entity(self, state: State, id: str, action: str, payload: t.Dict[str,
...
+
+            client_id, module_context, action, entity_id, current_status, jobs = res
+            if not action or not client_id or not entity_id or not jobs:
+                return
+
+            entity = core_get(entity_id)
+            if not entity:
+                return
+
+            on_action_function = self.gui._get_user_function(action)
```
Comment (FabienLelaquais) [inline src/taipy/gui_core/_context.py]: `on_action` is kind of confusing... (well, 'action' is)

File: src/taipy/gui_core/_context.py [hunk 3]
```diff
@@ -69,9 +81,22 @@ def __init__(self, gui: Gui) -> None:
         self.data_nodes_by_owner: t.Optional[t.Dict[t.Optional[str], DataNode]] = None
         self.scenario_configs: t.Optional[t.List[t.Tuple[str, str]]] = None
         self.jobs_list: t.Optional[t.List[Job]] = None
+        self.client_jobs_by_submission: t.Dict[
+            str,
+            t.Tuple[
```
Comment (FabienLelaquais) [inline src/taipy/gui_core/_context.py]: Maybe use a namedtuple here to better follow what's going on?

File: src/taipy/gui_core/_context.py [hunk 4]
```diff
@@ -271,11 +298,110 @@ def submit_entity(self, state: State, id: str, action: str, payload: t.Dict[str,
...
                 state.assign(_GuiCoreContext._SCENARIO_VIZ_ERROR_VAR, f"Error submitting entity. {e}")
 
+    def _get_submittable_status(self, jobs_ids: t.List[str]):
+        status = _SubmissionStatus.UNDEFINED
+        blocked = False
+        waiting = False
+        running = False
+        completed = False
+        for id in jobs_ids:
+            job = core_get(id)
```
Comment (FabienLelaquais) [inline src/taipy/gui_core/_context.py:344]: I'm a bit skeptic about the Core side performance of this... get() for each job...

**Review by FabienLelaquais — DISMISSED** [review]
👍👍👍

File: src/taipy/gui_core/_context.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui_core/_context.py:413]: something like ?

File: src/taipy/gui_core/_context.py (hunk 2 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui_core/_context.py]: ok

File: src/taipy/gui_core/_context.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui_core/_context.py:344]: maybe @jrobinAV can chip in

File: src/taipy/gui_core/_context.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui_core/_context.py]: really ?
this is read once and written twice

File: src/taipy/gui_core/_context.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui_core/_context.py:413]: expose a submission class from taipy ?
I would think it should come from core ?

File: src/taipy/gui_core/_context.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui_core/_context.py]: changed to a class

File: src/taipy/gui_core/_context.py [hunk 5]
```diff
@@ -271,11 +310,114 @@ def submit_entity(self, state: State, id: str, action: str, payload: t.Dict[str,
...
+                                job_ids,
+                            )
+                        self.scenario_status_callback(jobs[0].id if isinstance(jobs, list) else jobs.id)
+                    else:
+                        _warn(f"on_submission_change(): '{submission_cb}' is not a valid function.")
                 state.assign(_GuiCoreContext._SCENARIO_VIZ_ERROR_VAR, "")
             except Exception as e:
                 state.assign(_GuiCoreContext._SCENARIO_VIZ_ERROR_VAR, f"Error submitting entity. {e}")
 
+    def _get_submittable_status(self, jobs_ids: t.List[str]):
```
Comment (jrobinAV) [inline src/taipy/gui_core/_context.py]: We can add some return type: _SubmissionStatus

**Review by jrobinAV — COMMENTED** [review]
I cannot be sure the Submission status computation is correct just by reading the for loop. A bunch of unit tests would definitely help.

This method aside, We have a 700+ lines file without a single test. It is very difficult to read, to understand and to guarantee non regressions. We need to create some unit tests and eventually splitting the code into dedicated classes, modules, packages.

File: src/taipy/gui_core/viselements.json [hunk 6]
```diff
@@ -149,6 +149,12 @@
                         "type": "bool",
                         "default_value": "True",
                         "doc": "If False, the buttons to submit scenario sequences are not visible."
+                    },
+                    {
+                        "name": "on_submission_change",
+                        "type": "Callback",
+                        "doc": "The name of the function that is triggered when a submission status is changed.<br/><br/>All the parameters of that function are optional:\n<ul>\n<li>state (<code>State^</code>): the state instance.</li>\n<li>submittable (Submittable): the identity (usually a Scenario) that was submitted.</li>\n<li>details (dict): the details on this callback's invocation.<br/>\nThis dictionary has the following keys:\n<ul>\n<li>submission_status (str): the new status of the submission.</li>\n<li>job: the Job (if any) that is at the orgin of the submission status change.</li>\n</ul>",
```
Comment (FabienLelaquais) [inline src/taipy/gui_core/viselements.json]: "the identity"? You mean the entity?
Should we not list the possible statuses?

**Review by FabienLelaquais — DISMISSED** [review]
👌

**Review by FabienLelaquais — DISMISSED** [review]
👍

**Review by jrobinAV — APPROVED** [review]

**Review by FabienLelaquais — APPROVED** [review]
👌

**Pull request merged**